73-offline-attestation.md

```diff
@@ -0,0 +1,302 @@
+record and store it on the time-series persistent datastore. 
+- Two pieces of information: the name of a python module to be dynamically
+  imported (which will contain code used to interact with these new proposed
+stores) and the connection paramaters (for these new proposed stores) will be
+supplied by the user as two new parameters under `[cloud_verifier]` and
+`[registrar]` section: `persistent_store_import` and
+`persistent_store_connection_data`)
+- A new CLI interface - `keylime_attest` - will contact both the transparency
+  log and the time-series datastore, get a list of AIKs proven to be associated
+with an EK, and then call the same code used by the `verifier` (i.e.,
+`cloud_verifier_common.process_quote_response`) to perform a series of point in
+time attestation on all records retrieved from the persistent datastore.
+- An additional minor change: it is expected that the `verifier` will extract
+  the "TPM clock information" (i.e, "clock", "resetCount", "restartCount",
+"safe") and make it available as part of the `json_response` 
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- The `keylime_attest` CLI will call the attestation code used by the
+  `verifier` without any modification, and should be up to the user to write a
+more complex policy if he choses to do so.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+enhancement ecosystem.
+
+How will security be reviewed and by whom?
+-->
+
+- The key security aspect here is to convince ourselves (and others) that the
+  record generated by the `registrar` to indicate the association between EK
+and AIK is enough. Once this is done, offline attestation has basically the
+same level of security of the online attestation (which was already evaluated)
+as it uses the very same code base. 
+- While we do expect very little impact on KeyLime's scalability by adding this
+  capability, it is important to remember that we are constantly testing KL in
+a configuration with 5K nodes (with both MB and IMA simultaneously activated),
+and can provide experimental evidence to back this hypothesis.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+- The first PR will provide the "persistent datastore" plugin capability, to be
+  called from with both `registrar` and `verifier` code. It will include a
```

Comment:
It would be good to better understand the plugin architecture. Is this something that canonicalises materials into json and makes some sort of POST to the datastore.

Reply:
At this point, I can experimentally demonstrate that I can simply "jsonify" everything from within the `registrar`/`verifier` code (basically the `agent` and `json_response` python dictionaries) and push it into a datastore.

Reply:
I agree with OP, IMHO there is more value of defining the API (open_datastore(), close_datastore(), is_valid(), insert_line(), etc) and the security guarantees.

The json serialization sure is OK, but it is the API (and the parameters required) the one that will make sure that all the data will be stored also in future versions. If not there is a risk that one of those internal state objects that we are storing change, and make the attestation impossible (or worse, fix the internal representation for ever).

Reply:
@aplanas very well, I will add a proto-API to the enhancement text now.